stratumv2-lib/src/macro_message/setup_connection.rs

```diff
@@ -0,0 +1,188 @@
+pub mod macro_prelude {
+    pub use crate::error::{Error, Result};
+    pub use crate::frame::Frameable;
+    pub use crate::parse::{ByteParser, Deserializable, Serializable};
+    pub use crate::types::{MessageType, STR0_255};
+    pub use std::convert::TryInto;
+    pub use std::io;
+}
+
+/// Implemention of the requirements for a SetupConnection message for each
+/// sub protocol.
+#[macro_export]
+macro_rules! impl_setup_connection {
+    ($flags_type:ident) => {
+        use crate::macro_message::setup_connection::macro_prelude::*;
+
+        /// SetupConnection is the first message sent by a client on a new connection.
+        ///
+        /// The SetupConnection struct contains all the common fields for the
+        /// SetupConnection message for each Stratum V2 subprotocol.
+        ///
+        /// # Examples
+        ///
+        /// ```rust
+        /// use stratumv2_lib::mining;
+        /// use stratumv2_lib::job_negotiation;
+        ///
+        /// let mining_connection = mining::SetupConnection::new(
+        ///    2,
+        ///    2,
+        ///    mining::SetupConnectionFlags::REQUIRES_STANDARD_JOBS | mining::SetupConnectionFlags::REQUIRES_VERSION_ROLLING,
+        ///    "0.0.0.0",
+        ///    8545,
+        ///    "Bitmain",
+        ///    "S9i 13.5",
+        ///    "braiins-os-2018-09-22-1-hash",
+        ///    "some-device-uuid",
+        /// );
+        /// assert!(mining_connection.is_ok());
+        /// assert!(mining_connection.unwrap().flags.contains(
+        ///     mining::SetupConnectionFlags::REQUIRES_STANDARD_JOBS)
+        /// );
+        ///
+        /// let job_negotiation_connection = job_negotiation::SetupConnection::new(
+        ///    2,
+        ///    2,
+        ///    job_negotiation::SetupConnectionFlags::REQUIRES_ASYNC_JOB_MINING,
+        ///    "0.0.0.0",
+        ///    8545,
+        ///    "Bitmain",
+        ///    "S9i 13.5",
+        ///    "braiins-os-2018-09-22-1-hash",
+        ///    "some-device-uuid",
+        /// );
+        /// assert!(job_negotiation_connection.is_ok());
+        /// ```
+        #[derive(Debug, Clone)]
+        pub struct SetupConnection {
+            /// The minimum protocol version the client supports. (current default: 2)
```

Comment:
thanks! this makes sense. I played around with the `SetupConnection Enum` a bit, would it make sense to apply the `Frameable` trait to the `SetupConnection Enum` instead of the `SetupConnection Structs` in each sub protocol? 

In this case it allows the caller to create a `Network Message` from the `SetupConnection Enum` as they wouldn't be directly framing the `SetupConnection Structs` since it doesn't have the correct serialization structure (missing u8 enum protocol).

Let me know what you think, I'm happy to do that in a separate PR so we can merge this soon!

Reply:
After reviewing different approaches I came to the conclusion that I think the approach in #124 is more desirable. Here is my reasoning:

**Approach 1 (This approach):**

- Currently, if using the `SetupConnectionEnum`, the caller needs to create a new `mining::SetupConnection` by wrapping a `mining::SetupConnection` inside of a `SetupConnectionEnum::Mining`:

```
 let setup_conn = SetupConnectionEnum::Mining(
        mining::SetupConnection::new(
            2,
            2,
            mining::SetupConnectionFlags::REQUIRES_STANDARD_JOBS,
            "0.0.0.0",
            8545,
            "Bitmain",
            "S9i 13.5",
            "braiins-os-2018-09-22-1-hash",
            "some-uuid",
        )
        .unwrap(),
    );
```

- Calling `unframe` on a `Message` isn't able to deserialize to the `SetupConnectionEnum` since it doesn't implement `Frameable`. If we try to `unframe` directly to the `SetupConnection Struct`, the deserializer will not know how to handle the additional `u8` for the Protocol and will result in a `UnknownFlag` error.

**Approach 2 (#124):**

- Approach 2, defines the `Protocol enum` in the `SetupConnection macro message` as a crate only variable, so the caller will never have to deal with it since it's defined by the library. The caller just needs to initialize the SetupConnection message according to the module.

```
let setup_conn =  mining::SetupConnection::new(
        2,
        2,
        mining::SetupConnectionFlags::REQUIRES_STANDARD_JOBS,
        "0.0.0.0",
        8545,
        "Bitmain",
        "S9i 13.5",
        "braiins-os-2018-09-22-1-hash",
        "some-uuid",
)
.unwrap()
```

- Calling `unframe` on a `Message` will be able to handle the additional `Protocol` u8 without having to go through a intermediate `SetupConnection` enum

Interested to hear your thoughts!

Reply:
For approach (1) we could very easily implement `Frameable` for the `SetupConnection` enum. I intended to do that; I believe it was an oversight on my part to omit that.

For approach (2), I'm worried about the ergonomics of unframing one of the 4 `SetupConnection` message types. We need to know which concrete type to provide to `unframe`. Normally the way we would determine that is to peel the `Message` header off of the payload and switch on the `message_type` field. For most of our message types (aka, implementors of `Frameable`) there is exactly one Rust struct per `MessageType` variant. However, for `SetupConnection` there are 4 different possible structs. That makes it impossible to invoke `unframe` without first doing some additional parsing of `Message.payload` to find the protocol value.

I intended to address that discrepancy by using an enum as the top-level message for `SetupConnection` payloads. We would determine the underlying protocol type in the enum's `Deserializable` implementation, and we would just populate the appropriate variant with the fully-typed `SetupConnection` sub-message. This would keep the control flow of message receipt consistent despite the quirks of this message type. However, this does leak an implementation detail to the users of the library ("why are these messages bundled into an enum while everything else is its own top-level struct?"). I think either solution is fine, we're just choosing which quirk we want to deal with. Either
1. a consistent, but slightly confusing type interface, or
2. small amounts of special-case message handling logic, but a uniform type interface

Reply:
I like the benefits of both approaches but I don't like either trade off lol.

I'll merge this today once I've got a bit of time, but here's a proposal to simply the interface for the caller using the `SetupConnection Enum`:

```
impl SetupConnection {
    pub fn new_mining<T: Into<String>>(
        min_version: u16,
        max_version: u16,
        flags: mining::SetupConnectionFlags,
        endpoint_host: T,
        endpoint_port: u16,
        vendor: T,
        hardware_version: T,
        firmware: T,
        device_id: T,
    ) -> Result<SetupConnection> {
        Ok(SetupConnection::Mining(mining::SetupConnection::new(
            min_version,
            max_version,
            flags,
            endpoint_host,
            endpoint_port,
            vendor,
            hardware_version,
            firmware,
            device_id,
        )?))
    }

    pub fn new_job_negotation<T: Into<String>>(
        min_version: u16,
        max_version: u16,
        flags: job_negotiation::SetupConnectionFlags,
        endpoint_host: T,
        endpoint_port: u16,
        vendor: T,
        hardware_version: T,
        firmware: T,
        device_id: T,
    ) -> Result<SetupConnection> {
        Ok(SetupConnection::JobNegotiation(
            job_negotiation::SetupConnection::new(
                min_version,
                max_version,
                flags,
                endpoint_host,
                endpoint_port,
                vendor,
                hardware_version,
                firmware,
                device_id,
            )?,
        ))
    }
}
```

So I'm not a fan of the repeated function signature and also the pass through calls but at least it allows the caller to have a simple interface instead of wrapping the `SetupConnection struct` in the `SetupConnection enum`. I'll create this in another PR and propose it and also update the docs to encourage this use of creating and deserializing `SetupConnection` messages using the `enum`. If the caller wants to dig deep into the lib they can create the enum and struct separately but I would like to encourage them to take the path of least resistance!

Reply:
I think that's a good compromise. It's a little repetitive, but I agree that it's nicer for consumers.